File: Samples.Objc/Samples.ObjC/ViewControllers/CheckoutViewController/CheckoutViewController.m
//
//  CheckoutViewController.m
//  PayPalNativeCheckoutObjC
//
//  Created by Nguyen, The Nhat Minh on 3/9/21.
//  Copyright © 2020 PayPal. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "CheckoutViewController.h"
#import "AddItemViewController.h"
#import "AddItemCell.h"
#import "ItemCell.h"
#import "TotalCell.h"
#import "FetchAccessTokenEndpoint.h"
#import "CreateOrderEndpoint.h"
#import "AccessTokenResponse.h"
#import "CreateOrderResponse.h"
#import "UIViewController+Extension.h"
#import "PayPalAPI.h"
@import PayPalCheckout;

@interface CheckoutViewController ()
@property (nonatomic) UISegmentedControl *segmentedControl;
@property (nonatomic) UITableView *tableView;
@property (nonatomic) UIButton *checkoutButton;
@property (nonatomic) NSMutableArray<PPCPurchaseUnitItem *> *items;
@end

@implementation CheckoutViewController

// MARK: - Lifecycle

- (void)viewDidLoad {
  [super viewDidLoad];
  [self setupSampleItem];
  [self setupUI];
  [self setupTableView];
  [self setupConstraints];
}

// MARK: - PayPalCheckout

- (void)startCheckout {
  [PPCheckout
  startWithPresentingViewController:self
  createOrder:^(PPCCreateOrderAction *action) {
    switch (self.segmentedControl.selectedSegmentIndex) {
      case 0:
        // Checkout with order
        [self createOrderCallbackCreateOrderWithAction:action];
        break;
      case 1:
        // Checkout with orderID/ECToken/payToken
        [self createOrderCallbackSetOrderIDWithAction:action];
        break;
      default:
        break;
    }
  }
  onApprove:^(PPCApproval *approval) {
    [self onApproveCallbackWithApproval:approval];
  }
  onShippingChange: nil
  onCancel:^{
    [self onCancelCallback];
  }
  onError:^(PPCErrorInfo *errorInfo) {
    [self onErrorCallbackWithErrorInfo:errorInfo];
  }];
}

- (PPCOrderRequest*)createNewOrder {
  NSString *total = [self getTotal];
  NSString *itemTotal = [self getItemTotal];
  NSString *taxTotal = [self getTaxTotal];

  PPCUnitAmount *itemTotalUnitAmount = [[PPCUnitAmount alloc] initWithCurrencyCode:PPCCurrencyCodeUsd
                                                                             value:itemTotal];
  PPCUnitAmount *taxTotalUnitAmount = [[PPCUnitAmount alloc] initWithCurrencyCode:PPCCurrencyCodeUsd
                                                                            value:taxTotal];

  PPCPurchaseUnitBreakdown *purchaseUnitBreakdown = [[PPCPurchaseUnitBreakdown alloc] initWithItemTotal:itemTotalUnitAmount
                                                                                               shipping:nil
                                                                                               handling:nil
                                                                                               taxTotal:taxTotalUnitAmount
                                                                                              insurance:nil
                                                                                       shippingDiscount:nil
                                                                                               discount:nil];

  PPCPurchaseUnitAmount *purchaseUnitAmount = [[PPCPurchaseUnitAmount alloc] initWithCurrencyCode:PPCCurrencyCodeUsd
                                                                                            value:total
                                                                                        breakdown:purchaseUnitBreakdown];
  
  PPCPurchaseUnit *purchaseUnit = [[PPCPurchaseUnit alloc] initWithAmount:purchaseUnitAmount
                                                              referenceId:nil
                                                                    payee:nil
                                                       paymentInstruction:nil
                                                  purchaseUnitDescription:nil
                                                                 customId:nil
                                                                invoiceId:nil
                                                           softDescriptor:nil
                                                                    items:self.items
                                                                 shipping:nil];

  PPCOrderRequest *order = [[PPCOrderRequest alloc] initWithIntent:PPCOrderIntentAuthorize
                                                     purchaseUnits:@[purchaseUnit]
                                             processingInstruction:PPCOrderProcessingInstructionNone
                                                             payer:nil
                                                applicationContext:nil];
  return order;
}

/// createOrder callback:
/// This will be called when PayPalCheckout starts creating an order
/// Use this if you want PayPalCheckout to create an order and get an order ID for you internally via PayPal Orders API
- (void)createOrderCallbackCreateOrderWithAction:(PPCCreateOrderAction *)action {
  PPCOrderRequest *order = [self createNewOrder];
  [action createWithOrder:order completion:^(NSString *orderId) {
    NSLog(@"Order created with orderId: %@", orderId);
  }];
}

/// createOrder callback:
/// This will be called when PayPalCheckout starts creating an order
/// Use this if you want to create an order an get an order ID with PayPal Orders API by yourself
- (void)createOrderCallbackSetOrderIDWithAction:(PPCCreateOrderAction *)action {
  PPCOrderRequest *order = [self createNewOrder];
  
  NSString *clientId = [PayPalAPI.shared clientId];
  FetchAccessTokenRequest *tokenRequest = [[FetchAccessTokenRequest alloc] initWith:clientId];
  [PayPalAPI.shared fetchAccessToken:tokenRequest completion:^(NSData *data, NSError *error) {
    if (data == nil) {
      NSLog(@"Fetch access token failed with no data.");
      return;
    }
    
    AccessTokenResponse *tokenResponse = [[AccessTokenResponse alloc] initWithData:data];
    CreateOrderRequest *createOrderRequest = [[CreateOrderRequest alloc] initWithOrder:order accessToken:tokenResponse.accessToken];
    
    [PayPalAPI.shared createOrder:createOrderRequest completion:^(NSData *data, NSError *error) {
      if (data == nil) {
        NSLog(@"Create order failed with no data.");
        return;
      }
      
      CreateOrderResponse *orderResponse = [[CreateOrderResponse alloc] initWithData:data];
      [action setWithOrderId:orderResponse.orderId];
    }];
  }];
}

/// onApprove callback: This will be called when checkout with PayPalCheckout is completed, you will need to handle authorizing or capturing the funds in this callback
- (void)onApproveCallbackWithApproval:(PPCApproval *)approval {
  if (approval.data.intent == PPCOrderIntentAuthorize) {
    [approval.actions authorizeOnComplete:^(PPCAuthorizeActionSuccess *success, NSError *error) {
      if (error) {
        NSLog(@"Fail to authorize order with error %@", error.localizedDescription);
      } else if (success) {
        NSLog(@"Authorize order successfully");
      } else {
        NSLog(@"Authorize order: No error and no success response");
      }
    }];
  } else if (approval.data.intent == PPCOrderIntentCapture) {
    [approval.actions captureOnComplete:^(PPCCaptureActionSuccess *success, NSError *error) {
      if (error) {
        NSLog(@"Fail to capture order with error %@", error.localizedDescription);
      } else if (success) {
        NSLog(@"Capture order successfully");
      } else {
        NSLog(@"Capture order: No error and no success response");
      }
    }];
  }
}

/// onCancel callback: This will be called when users cancel checkout
- (void)onCancelCallback {
  NSLog(@"Checkout cancelled");
}

/// onError callback: This will be call when an error occurs in the checkout session
- (void)onErrorCallbackWithErrorInfo:(PPCErrorInfo *)errorInfo {
  NSLog(@"Checkout failed with error: %@", errorInfo.error.localizedDescription);
}

// MARK: - UI

- (void)setupTableView {
  [self.tableView registerClass:[AddItemCell class] forCellReuseIdentifier:@"AddItemCell"];
  self.tableView.delegate = self;
  self.tableView.dataSource = self;
}

- (void)setupUI {
  self.view.backgroundColor = [UIColor whiteColor];
  
  NSArray *options = @[@"Order", @"ECToken"];
  self.segmentedControl = [[UISegmentedControl alloc] initWithItems:options];
  [self.segmentedControl addTarget:self action:@selector(segmentChanged:) forControlEvents:UIControlEventValueChanged];
  self.segmentedControl.selectedSegmentIndex = 0;
  self.segmentedControl.translatesAutoresizingMaskIntoConstraints = false;
  
  self.tableView = [[UITableView alloc] init];
  self.tableView.translatesAutoresizingMaskIntoConstraints = false;

  self.checkoutButton = [UIViewController buttonWithColor:[UIColor systemBlueColor] textColor:[UIColor whiteColor] title:@"Checkout with order" font:[UIFont boldSystemFontOfSize:16]];
  [self.checkoutButton addTarget:self action:@selector(didTapCheckout) forControlEvents:UIControlEventTouchUpInside];
  self.checkoutButton.translatesAutoresizingMaskIntoConstraints = false;

  [self.view addSubview:self.segmentedControl];
  [self.view addSubview:self.tableView];
  [self.view addSubview: self.checkoutButton];
  
}

- (void)setupConstraints {
  [[self.segmentedControl.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:16] setActive:true];
  [[self.segmentedControl.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:16] setActive:true];
  [[self.segmentedControl.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-16] setActive:true];
  [[self.segmentedControl.heightAnchor constraintEqualToConstant:32] setActive:true];
  [[self.segmentedControl.bottomAnchor constraintEqualToAnchor:self.tableView.topAnchor constant:-24] setActive:true];
  
  [[self.tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor] setActive:true];
  [[self.tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor] setActive:true];
  [[self.tableView.bottomAnchor constraintEqualToAnchor:self.checkoutButton.topAnchor constant:16] setActive:true];
  
  [[self.checkoutButton.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:16] setActive:true];
  [[self.checkoutButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-16] setActive:true];
  [[self.checkoutButton.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-16] setActive:true];
  [[self.checkoutButton.heightAnchor constraintEqualToConstant:40] setActive:true];
}

- (void)segmentChanged:(UISegmentedControl *)segment {
  switch (segment.selectedSegmentIndex) {
    case 0:
      self.checkoutButton.hidden = false;
      [self.checkoutButton setTitle:@"Checkout with order" forState:UIControlStateNormal];
      break;
    case 1:
      self.checkoutButton.hidden = false;
      [self.checkoutButton setTitle:@"Checkout with ECToken" forState:UIControlStateNormal];
      break;
    case 2:
      self.checkoutButton.hidden = true;
      break;
    default:
      break;
  }
}

- (void)didTapCheckout {
  if (self.items.count == 0) {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Can't checkout"
                                                                   message:@"Please add at least 1 item"
                                                            preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK"
                                                     style:UIAlertActionStyleDefault
                                                   handler:^(UIAlertAction *_){}];
    [alert addAction:action];
    [self presentViewController:alert animated:true completion:^{}];
    return;
  }

  [self startCheckout];
}

// MARK: - Utility

- (void)setupSampleItem {
  PPCPurchaseUnitItem *item = [[PPCPurchaseUnitItem alloc] initWithName:@"Sample item"
                                                             unitAmount:[[PPCUnitAmount alloc] initWithCurrencyCode:PPCCurrencyCodeUsd value:@"5.99"]
                                                               quantity:@"1"
                                                                    tax:[[PPCPurchaseUnitTax alloc] initWithCurrencyCode:PPCCurrencyCodeUsd value:@"0.59"]
                                                        itemDescription:nil
                                                                    sku:nil
                                                               category:PPCPurchaseUnitCategoryPhysicalGoods];
  self.items = [[NSMutableArray alloc] init];
  [self.items addObject:item];
}

- (NSString *)getTotal {
  double total = 0;
  for (PPCPurchaseUnitItem* item in self.items) {
    double price = [item.unitAmount.value doubleValue];
    double tax = [item.tax.value doubleValue];
    double quantity = [item.quantity doubleValue];
    total += (price + tax) * quantity;
  }
  return [[NSNumber numberWithDouble:total] stringValue];
}

- (NSString *)getItemTotal {
  double total = 0;
  for (PPCPurchaseUnitItem* item in self.items) {
    double price = [item.unitAmount.value doubleValue];
    double quantity = [item.quantity doubleValue];
    total += price * quantity;
  }
  return [[NSNumber numberWithDouble:total] stringValue];
}

- (NSString *)getTaxTotal {
  double total = 0;
  for (PPCPurchaseUnitItem* item in self.items) {
    double tax = [item.tax.value doubleValue];
    double quantity = [item.quantity doubleValue];
    total += tax * quantity;
  }
  return [[NSNumber numberWithDouble:total] stringValue];
}

// MARK: - UITableViewDelegate and UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 2;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
  switch (section) {
    case 0:
      return @"Items";
    case 1:
      return @"Cart total";
    default:
      return @"";
  }
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
  if (indexPath.section == 0) {
    if (indexPath.row >= self.items.count) {
      AddItemCell *cell = [[AddItemCell alloc] init];
      return cell;
    } else {
      ItemCell *cell = [[ItemCell alloc] initWithItem:self.items[indexPath.row] andQuantity:self.items[indexPath.row].quantity];
      return cell;
    }
  } else {
    TotalCell *cell = [[TotalCell alloc] initWithTotal:[self getTotal] subtotal:[self getItemTotal] tax:[self getTaxTotal]];
    return cell;
  }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
  if (indexPath.section == 0 && indexPath.row >= self.items.count) {
    return 48;
  }
  return UITableViewAutomaticDimension;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  if (section == 0) {
    return self.items.count + 1;
  } else {
    return 1;
  }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  if (indexPath.section == 0) {
    if (indexPath.row >= self.items.count) {
      AddItemViewController *vc = [[AddItemViewController alloc] init];
      vc.delegate = self;
      vc.index = indexPath.row;
      [self presentViewController:vc animated:true completion:^{}];
    } else {
      AddItemViewController *vc = [[AddItemViewController alloc] initWithItem:self.items[indexPath.row]];
      vc.delegate = self;
      vc.index = indexPath.row;
      [self presentViewController:vc animated:true completion:^{}];
    }
  }
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
  return indexPath.section == 0 && indexPath.row < self.items.count;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
  if (editingStyle == UITableViewCellEditingStyleDelete) {
    [self.items removeObjectAtIndex:indexPath.row];
    [tableView performBatchUpdates:^{
      [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } completion:^(BOOL _) {
      [tableView reloadData];
    }];
  }
}

// MARK: - AddItemViewControllerDelegate
- (void)didTapSaveWithItem:(PPCPurchaseUnitItem *)item at:(NSInteger)index {
  if (index < self.items.count) {
    self.items[index] = item;
  } else {
    [self.items addObject:item];
  }
  [self.tableView reloadData];
}

@end

File: Samples.Swift/Samples.Swift/ViewControllers/CheckoutViewController/CheckoutViewController.swift
//
//  CheckoutViewController.swift
//  paypalcheckout-iOS-swift-sample
//
//  Created by Jax DesMarais-Leder on 3/10/21.
//  Copyright © 2021 PayPal. All rights reserved.
//

import UIKit
import PayPalCheckout

class CheckoutViewController: UIViewController, AddItemViewControllerDelegate {

  var items = [PurchaseUnit.Item]()

  lazy var tableView: UITableView = {
    let tableView = UITableView()
    tableView.register(AddItemCell.self, forCellReuseIdentifier: "AddItemCell")
    tableView.register(ItemCell.self, forCellReuseIdentifier: "ItemCell")
    tableView.register(TotalCell.self, forCellReuseIdentifier: "TotalCell")
    tableView.delegate = self
    tableView.dataSource = self
    tableView.translatesAutoresizingMaskIntoConstraints = false
    return tableView
  }()

  lazy var checkoutButton: UIButton = {
    let checkoutButton = UIButton()
    checkoutButton.setTitle("Checkout with Order", for: .normal)
    checkoutButton.layer.cornerRadius = 8
    checkoutButton.backgroundColor = .systemBlue
    checkoutButton.tintColor = .white
    checkoutButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
    checkoutButton.addTarget(self, action: #selector(tapCheckout), for: .touchUpInside)
    checkoutButton.translatesAutoresizingMaskIntoConstraints = false
    return checkoutButton
  }()

  lazy var paypalButton: PayPalButton = {
    let paypalButton = PayPalButton(size: .expanded)
    paypalButton.addTarget(self, action: #selector(tapCheckout), for: .touchUpInside)
    paypalButton.translatesAutoresizingMaskIntoConstraints = false
    return paypalButton
  }()

  lazy var checkoutFlowOption: UISegmentedControl = {
    let checkoutFlowOptions = ["Order", "ECToken", "Payment Button"]
    let checkoutFlowOption = UISegmentedControl(items: checkoutFlowOptions)

    checkoutFlowOption.addTarget(
      self,
      action: #selector(checkoutFlowOptionChanged),
      for: .valueChanged
    )

    checkoutFlowOption.selectedSegmentIndex = 0
    checkoutFlowOption.translatesAutoresizingMaskIntoConstraints = false
    return checkoutFlowOption
  }()


  override func viewDidLoad() {
    super.viewDidLoad()
    paypalButton.isHidden = true

    setupSampleItem()
    configure()
  }

  @objc
  func tapCheckout() {
    if items.count == 0 {
      let alertController = UIAlertController(title: "Cannot checkout", message: "Please add at least 1 item to the cart", preferredStyle: .alert)
      alertController.addAction(UIAlertAction(title: "Ok", style: .default))
      self.present(alertController, animated: true)
      return
    }

    startNativeCheckout()
  }

  func startNativeCheckout() {
    let order = createNewOrder()

    switch self.checkoutFlowOption.selectedSegmentIndex {
    /// Checkout with a cart created and let the SDK handle passing in the order ID
    case 0:
      Checkout.start(
        createOrder: { action in
          action.create(order: order) { orderId in
            if orderId == nil {
              print("There was an error with the format of your order object")
            }
            else {
              print("Order created with order ID \(String(describing: orderId))")
            }
          }
        },
        onApprove: { approval in
          self.processOrderActions(with: approval)
        },
        onCancel: {
          print("Checkout cancelled")
        },
        onError: { errorInfo in
          print("Checkout failed with error info \(errorInfo.error.localizedDescription)")
        }
      )

    /// Request an ECToken/orderID/payToken with PayPal Orders API,
    /// then checkout with `ECToken`/`orderID`/`payToken`
    case 1:
      Checkout.start(createOrder: { action in
        let tokenRequest = AccessTokenRequest(clientId: PayPal.clientId)

        PayPal.shared.request(on: .fetchAccessToken(tokenRequest)) { data, error in

          guard
            let data = data,
            let tokenResponse = try? PayPal.jsonDecoder.decode(AccessTokenResponse.self, from: data)
          else {
            print("Fetch access token failed with no data")
            return
          }

          PayPal.shared.setAccessToken(tokenResponse.accessToken)

          PayPal.shared.request(on: .createOrder(order)) {
            data, error in
            guard
              let data = data,
              let createOrderResponse = try? PayPal.jsonDecoder.decode(CreateOrderResponse.self, from: data)
            else {
              print("Create order failed with no data")
              return
            }

            action.set(orderId: createOrderResponse.id)
          }
        }
      },
      onApprove: { approval in
        self.processOrderActions(with: approval)
      },
      onCancel: {
        print("Checkout cancelled")
      },
      onError: { errorInfo in
        print("Checkout failed with error info \(errorInfo.error.localizedDescription)")
      }
      )

    /// Checkout with payment buttons
    case 2:
      Checkout.setCreateOrderCallback { action in
        action.create(order: order) { orderId in
          if orderId == nil {
            print("There was an error with the format of your order object")
          }
          else {
            print("Order created with order ID \(String(describing: orderId))")
          }
        }
      }

      Checkout.setOnApproveCallback { approval in
        self.processOrderActions(with: approval)
      }

      Checkout.setOnCancelCallback {
        print("Checkout cancelled")
      }

      Checkout.setOnErrorCallback { errorInfo in
        print("Checkout failed with error info \(errorInfo.error.localizedDescription)")
      }

    default:
      break
    }
  }

  func createNewOrder() -> OrderRequest {
    let purchaseUnit: PurchaseUnit = PurchaseUnit(
      amount: PurchaseUnit.Amount(
        currencyCode: .usd,
        value: String(self.getTotal()),
        breakdown: PurchaseUnit.Breakdown(
          itemTotal: UnitAmount(
            currencyCode: .usd,
            value: String(self.getItemTotal())
          ),
          taxTotal: UnitAmount(
            currencyCode: .usd,
            value: String(self.getTaxTotal())
          )
        )
      ),
      items: items
    )


    let order: OrderRequest = OrderRequest(
      intent: .authorize,
      purchaseUnits: [purchaseUnit]
    )

    return order
  }

  func setupSampleItem() {
    let item: PurchaseUnit.Item = PurchaseUnit.Item(
      name: "Sample item",
      unitAmount: UnitAmount(
        currencyCode: .usd, value: "5.99"
      ),
      quantity: "1",
      tax: PurchaseUnit.Tax(
        currencyCode: .usd,
        value: "1.23"
      )
    )

    items.append(item)
  }

  func configure() {
    view.backgroundColor = .white

    view.addSubview(checkoutFlowOption)
    view.addSubview(tableView)
    view.addSubview(checkoutButton)
    view.addSubview(paypalButton)

    NSLayoutConstraint.activate(getConstraints())
  }

  func getConstraints() -> [NSLayoutConstraint] {
    [
      checkoutFlowOption.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 10),
      checkoutFlowOption.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
      checkoutFlowOption.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
      checkoutFlowOption.heightAnchor.constraint(equalToConstant: 32),
      checkoutFlowOption.bottomAnchor.constraint(equalTo: tableView.topAnchor, constant: -24),

      tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
      tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
      tableView.bottomAnchor.constraint(equalTo: checkoutButton.topAnchor, constant: 20),

      checkoutButton.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
      checkoutButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
      checkoutButton.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -10),
      checkoutButton.heightAnchor.constraint(equalToConstant: 40),

      paypalButton.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
      paypalButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
      paypalButton.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -10),
      paypalButton.heightAnchor.constraint(equalToConstant: 40),
    ]
  }

  @objc
  func checkoutFlowOptionChanged(_ sender: UISegmentedControl) {
    switch sender.selectedSegmentIndex {
    case 0:
      checkoutButton.isHidden = false
      paypalButton.isHidden = true
      checkoutButton.setTitle("Checkout with Order", for: .normal)

    case 1:
      checkoutButton.isHidden = false
      paypalButton.isHidden = true
      checkoutButton.setTitle("Checkout with ECToken", for: .normal)

    case 2:
      checkoutButton.isHidden = true
      paypalButton.isHidden = false

    default:
      break
    }
  }

  private func processOrderActions(with approval: Approval) {
    switch approval.data.intent {
    case "AUTHORIZE":
      approval.actions.authorize { success, error in
        if let error = error {
          self.errorResultAction(errorMessage: error.localizedDescription)
        }
        else if let success = success {
          self.successResultAction(orderData: success.data.orderData, orderId: success.data.id)
        }
        else {
          print("Unhandled State: No Error and No Success Response")
        }
      }

    case "CAPTURE", "SALE":
      approval.actions.capture { success, error in
        if let error = error {
          self.errorResultAction(errorMessage: error.localizedDescription)
        }
        else if let success = success {
          self.successResultAction(orderData: success.data.orderData, orderId: success.data.id)
        }
        else {
          print("Unhandled State: No Error and No Success Response")
        }
      }

    default:
      break
    }
  }

  func successResultAction(orderData: [String: Any], orderId: String) {
    print("Successful Response Data: \(orderData)")

    let resultViewController = ResultViewController()
    resultViewController.resultLabel.text = "Success 😎"
    resultViewController.messageLabel.text = "Your order number is \(orderId)"
    resultViewController.modalPresentationStyle = .fullScreen

    self.present(resultViewController, animated: true)
  }

  func errorResultAction(errorMessage: String) {
      print("Capture Action Error Response: \(errorMessage)")

      let resultViewController = ResultViewController()
      resultViewController.resultLabel.text = "Error 😕"
      resultViewController.messageLabel.text = "Your order failed with \(errorMessage)"
      resultViewController.modalPresentationStyle = .fullScreen

    self.present(resultViewController, animated: true)
  }

  func getItemTotal() -> String {
    let total = items.reduce(0) { (runningTotal, item) -> Double in
       let unitAmount = Double(item.unitAmount.value ?? "0") ?? 0
       let quantity = Double(item.quantity) ?? 0
       return runningTotal + (quantity * unitAmount)
     }

    return String(format: "%.2f", total)
  }

  func getTaxTotal() -> String {
    let total = items.reduce(0) { (runningTotal, item) -> Double in
      let taxPrice = Double(item.tax?.value ?? "0") ?? 0
      let quantity = Double(item.quantity) ?? 0
      return runningTotal + (taxPrice * quantity)
    }
    
    return String(format: "%.2f", total)
  }

  func getTotal() -> String {
    let total = items.reduce(0) { (runningTotal, item) -> Double in
      let unitPrice = Double(item.unitAmount.value ?? "0") ?? 0
      let taxPrice = Double(item.tax?.value ?? "0") ?? 0
      let quantity = Double(item.quantity) ?? 0
      return runningTotal + ((unitPrice * quantity) + (taxPrice * quantity))
    }

    return String(format: "%.2f", total)
  }

  // MARK: AddItemViewControllerDelegate
  func didTapSaveWithItem(item: PurchaseUnit.Item, at index: Int) {
    if index < items.count {
      items[index] = item
    }
    else {
      items.append(item)
    }
    tableView.reloadData()
  }

  deinit { }
}

extension CheckoutViewController: UITableViewDelegate, UITableViewDataSource {

  func numberOfSections(in tableView: UITableView) -> Int {
      return 2
  }

  func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
    switch section {
    case 0:
      return "Items"

    case 1:
      return "Cart total"

    default:
      return nil
    }
  }

  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    if indexPath.section == 0 {
      if indexPath.row >= items.count {
        guard
          let cell = tableView.dequeueReusableCell(withIdentifier: "AddItemCell", for: indexPath) as? AddItemCell
          else { return UITableViewCell() }
        return cell
      }
      else {
        guard
          let cell = tableView.dequeueReusableCell(withIdentifier: "ItemCell", for: indexPath) as? ItemCell
          else { return UITableViewCell() }
        cell.setupUIWithItem(withItem: items[indexPath.row])
        return cell
      }
    }
    else {
      guard
        let cell = tableView.dequeueReusableCell(withIdentifier: "TotalCell", for: indexPath) as? TotalCell
        else { return UITableViewCell() }
      cell.setupUIWithTotal(withSubtotal: "$\(getItemTotal())", withTax: "$\(getTaxTotal())", withTotal: "$\(getTotal())")
      return cell
    }
  }

  func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
    if indexPath.section == 0 && indexPath.row >= items.count {
      return 48
    } else {
      return UITableView.automaticDimension
    }
  }

  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return section == 0 ? items.count + 1 : 1
  }

  func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    guard indexPath.section == 0 else { return }
    if indexPath.row >= items.count {
      let addItemViewController = AddItemViewController()
      addItemViewController.modalPresentationStyle = .overFullScreen
      addItemViewController.delegate = self
      addItemViewController.index = indexPath.row
      present(addItemViewController, animated: true)
      tableView.deselectRow(at: indexPath, animated: true)
    }
    else {
      let addItemViewController = AddItemViewController(item: items[indexPath.row])

      addItemViewController.modalPresentationStyle = .overFullScreen
      addItemViewController.delegate = self
      addItemViewController.index = indexPath.row

      present(addItemViewController, animated: true)
      tableView.deselectRow(at: indexPath, animated: true)
    }
  }

  func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
    return indexPath.section == 0 && indexPath.row < self.items.count
  }

  func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
    if editingStyle == .delete {
      self.items.remove(at: indexPath.row)
      tableView.performBatchUpdates {
        tableView.deleteRows(at: [indexPath], with: .fade)
      } completion: { (Bool) in
        tableView.reloadData()
      }
    }
  }
}
